<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Table S1.xlsx</title>

<script src="site_libs/header-attrs-2.30/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>
<link href="site_libs/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet" />
<script src="site_libs/htmlwidgets-1.6.4/htmlwidgets.js"></script>
<link href="site_libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet" />
<script src="site_libs/datatables-binding-0.34.0/datatables.js"></script>
<link href="site_libs/dt-core-1.13.6/css/jquery.dataTables.min.css" rel="stylesheet" />
<link href="site_libs/dt-core-1.13.6/css/jquery.dataTables.extra.css" rel="stylesheet" />
<script src="site_libs/dt-core-1.13.6/js/jquery.dataTables.min.js"></script>
<script src="site_libs/jszip-1.13.6/jszip.min.js"></script>
<link href="site_libs/dt-ext-buttons-1.13.6/css/buttons.dataTables.min.css" rel="stylesheet" />
<script src="site_libs/dt-ext-buttons-1.13.6/js/dataTables.buttons.min.js"></script>
<script src="site_libs/dt-ext-buttons-1.13.6/js/buttons.html5.min.js"></script>
<script src="site_libs/dt-ext-buttons-1.13.6/js/buttons.colVis.min.js"></script>
<script src="site_libs/dt-ext-buttons-1.13.6/js/buttons.print.min.js"></script>
<link href="site_libs/dt-ext-scroller-1.13.6/css/scroller.dataTables.min.css" rel="stylesheet" />
<script src="site_libs/dt-ext-scroller-1.13.6/js/dataTables.scroller.min.js"></script>
<link href="site_libs/dt-plugin-searchhighlight-1.13.6/source.css" rel="stylesheet" />
<script src="site_libs/dt-plugin-searchhighlight-1.13.6/jquery.highlight.js"></script>
<script src="site_libs/dt-plugin-searchhighlight-1.13.6/source.min.js"></script>
<link href="site_libs/crosstalk-1.2.2/css/crosstalk.min.css" rel="stylesheet" />
<script src="site_libs/crosstalk-1.2.2/js/crosstalk.min.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Supplementary</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="umap.html">UMAP</a>
</li>
<li>
  <a href="figures.html">Figures</a>
</li>
<li>
  <a href="tables.html">Tables</a>
</li>
<li>
  <a href="scripts.html">Scripts</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Table S1.xlsx</h1>

</div>


<p>Dataset-level metadata for the sequencing datasets used in this
thesis.</p>
<p><strong>Mention in thesis (for traceability):</strong></p>
<blockquote>
<p>Dataset-level metadata, including NCBI accessions, where available,
treatment and extraction protocols, read configuration, and platform
details, are provided in Supplementary Table S1.</p>
<p>Bulk RNA-seq could detect strong whole-worm transcriptional
responses, but it reported an average signal from mixed cell types and
could obscure lineage-restricted changes, particularly when the relevant
populations were rare (X. Li and Wang 2021). Thus, in the subsequent
stage of the study, the impact of Smed ELAC2 silencing was examined at
the single-cell resolution. In this analysis, datasets from WT, GFP
mock, and Smed ELAC2 KD worms were used, encompassing samples collected
at early stages of regeneration (0, 16, 24, and 72 hpa) (Supplementary
Table S1). The 72 hpa time point corresponded to one of the bulk RNA-seq
sampling points (3 dpa). The generation of single-cell resolution gene
expression data was intended to be subsequently integrated with
information on sncRNA abundance and predicted functions, in order to
gain insight into their roles in regulating cell lineages.
Pre-processing included read mapping, cell and feature QC, mitochondrial
content filtering, doublet control, construction of 4 per-time point
separate Seurat objects, containing all 3 conditions, and initial
clustering. These steps were completed by Dr Małgorzata
Marszałek-Zeńczak, and all analyses reported below start from these
pre-processed objects (Section 3.9).</p>
</blockquote>
<ul>
<li><a href="tables/Table%20S1.xlsx" download>Download original
XLSX</a></li>
</ul>
<h2>Sheet: Sheet1</h2>
<div class="datatables html-widget html-fill-item" id="htmlwidget-9f4b2824c0d09522c9c6" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-9f4b2824c0d09522c9c6">{"x":{"filter":"none","vertical":false,"extensions":["Buttons","Scroller"],"data":[["PRJNA1243063","PRJNA1243063","PRJNA1246370","PRJNA1246370","PRJNA1246370","PRJNA1246370","N/A","N/A"],["Gene expression profiling of regenerating S. mediterranea (ELAC2 KD, GFP Mock, WT) at 3 and 5 dpa - replicates 1-3","Small RNA profiling of regenerating S. mediterranea (ELAC2 KD, GFP Mock, WT) at 3 and 5 dpa - replicates 1-3","Illumina short-read sequencing of rRNA-depleted RNA composite sample from WT S. mediterranea across multiple regeneration time points","PacBio Iso-Seq long-read sequencing of rRNA-depleted RNA composite sample from WT S. mediterranea across multiple regeneration time points","PacBio Iso-Seq long-read sequencing of total RNA from amputated WT S. mediterranea at 0 hpa","PacBio Iso-Seq long-read sequencing of rRNA-depleted RNA from amputated WT S. mediterranea at 0 hpa","Single-cell gene expression profiling of regenerating S. mediterranea (ELAC2 KD, GFP Mock, WT) across 0, 16, 24, 72 hpa with scDASH rRNA depletion (3′ GEX) – replicate 1","Small RNA profiling of FACS-sorted neoblast (D gate) and non-neoblast (non-D gate) fractions from regenerating S. mediterranea (ELAC2 KD, GFP Mock, WT) at 3 dpa – replicates 1–3"],["Schmidtea mediterranea","Schmidtea mediterranea","Schmidtea mediterranea","Schmidtea mediterranea","Schmidtea mediterranea","Schmidtea mediterranea","Schmidtea mediterranea","Schmidtea mediterranea"],["Expression profiling by high throughput sequencing","Small RNA profiling by high throughput sequencing","Expression profiling by high throughput sequencing","Expression profiling by high throughput, long-read sequencing","Expression profiling by high throughput, long-read sequencing","Expression profiling by high throughput, long-read sequencing","Expression profiling by high throughput sequencing (single-cell RNA-seq)","Small RNA profiling by low throughput sequencing"],["This series comprises samples collected at 3 and 5 days post-amputation. Three biological replicates per condition (ELAC2 KD, GFP Mock, WT) were processed in 2 batches.","This series comprises samples collected at 3 and 5 days post-amputation. Three biological replicates per condition (ELAC2 KD, GFP Mock, WT) were processed in 1 batch.","A single composite sample generated from a mixture of 8 time points (0, 1, 3, 6, 9, 12, 24, 48 hours post-amputation) plus one non-regenerating sample, representing WT specimens.","A single composite sample generated from a mixture of 8 time points (0, 1, 3, 6, 9, 12, 24, 48 hours post-amputation) plus one non-regenerating sample, representing WT specimens.","A single sample from WT worms at 0 hours post-amputation","A single sample from WT worms at 0 hours post-amputation with rRNA depletion applied prior to library preparation.","Single-cell 3′ GEX profiling from whole animals collected at 4 time points post-amputation (0, 16, 24, 72 hpa) in 3 conditions (WT, ELAC2 KD, GFP mock), with 1 biological replicate per condition-time point combination (12 libraries total); rRNA depletion was implemented at the cDNA stage using scDASH to improve usable library complexity","Bulk small RNA-seq from FACS-sorted cell fractions at 3 dpa; 3 biological replicates per condition (ELAC2 KD, GFP Mock, WT) and 2 cell populations (neoblast-enriched D vs non-neoblast non-D), yielding 18 RNA samples / libraries total"],["WT sample: Amputation-induced regeneration (no additional treatment).\r\nGFP sample: RNAi mediated gene knockdown mock experiments were performed by the introduction of artificial dsRNA synthesized against GFP using AmpliScribe T7-Flash Transcription Kit – Epicentre #ASF3507, via microinjections into the worms. Amputation-induced regeneration.\r\nELAC sample: RNAi mediated gene knockdown experiments were performed by the introduction of artificial dsRNA synthesized against Smed-ELAC2 using AmpliScribe T7-Flash Transcription Kit – Epicentre #ASF3507, via microinjections into the worms. Amputation-induced regeneration.","WT sample: Amputation-induced regeneration (no additional treatment).\r\nGFP sample: RNAi mediated gene knockdown mock experiments were performed by the introduction of artificial dsRNA synthesized against GFP using AmpliScribe T7-Flash Transcription Kit – Epicentre #ASF3507, via microinjections into the worms. Amputation-induced regeneration.\r\nELAC sample: RNAi mediated gene knockdown experiments were performed by the introduction of artificial dsRNA synthesized against Smed-ELAC2 using AmpliScribe T7-Flash Transcription Kit – Epicentre #ASF3507, via microinjections into the worms. Amputation-induced regeneration.","No additional treatment; represents baseline and regenerating conditions","No treatment; represents baseline and regenerative state","No treatment; amputated and harvested immediately (0 hpa)","No treatment; amputated and harvested immediately (0 hpa)","WT sample: Amputation-induced regeneration (no additional treatment).\r\nGFP sample: RNAi mediated gene knockdown mock experiments were performed by the introduction of artificial dsRNA synthesized against GFP using AmpliScribe T7-Flash Transcription Kit – Epicentre #ASF3507, via microinjections into the worms. Amputation-induced regeneration.\r\nELAC sample: RNAi mediated gene knockdown experiments were performed by the introduction of artificial dsRNA synthesized against Smed-ELAC2 using AmpliScribe T7-Flash Transcription Kit – Epicentre #ASF3507, via microinjections into the worms. Amputation-induced regeneration.","WT sample: Amputation-induced regeneration (no additional treatment).\r\nGFP sample: RNAi mediated gene knockdown mock experiments were performed by the introduction of artificial dsRNA synthesized against GFP using AmpliScribe T7-Flash Transcription Kit – Epicentre #ASF3507, via microinjections into the worms. Amputation-induced regeneration.\r\nELAC sample: RNAi mediated gene knockdown experiments were performed by the introduction of artificial dsRNA synthesized against Smed-ELAC2 using AmpliScribe T7-Flash Transcription Kit – Epicentre #ASF3507, via microinjections into the worms. Amputation-induced regeneration. \r\nDissociated cells were stained with DRAQ5 and Calcein AM and sorted into neoblasts and non-neoblasts fractions"],["Standard planarian culture conditions","Standard planarian culture conditions","Standard planarian culture conditions","Standard planarian culture conditions","Standard planarian culture conditions","Standard planarian culture conditions","Standard planarian culture conditions","Standard planarian culture conditions"],["Long RNA fraction (&gt;200 nt) – rRNA depletion was unsuccessful","Short RNA fraction (&lt;200nt)","rRNA-depleted RNA","rRNA-depleted RNA","Total RNA","rRNA depleted RNA","Single-cell mRNA captured as 3′ cDNA (GEX)","Short RNA fraction (&lt;200nt) (from RNA isolated from FACS-sorted cells)"],["Worms were flash-frozen in liquid nitrogen. Total RNA was isolated using the mirVana™ miRNA Isolation Kit following the manufacturer’s protocol, then further fractionated into long (&gt;200 nt) and short (&lt; 200 nt) RNA fractions using the same kit. Long RNA fractions were then treated with TURBO™ DNase as per the provided instructions. The concentration of the purified RNA samples was quantified using a Qubit 4 Fluorometer and the Qubit™ RNA Broad Range (BR) Assay Kit. RNA quality was assessed using the Agilent TapeStation system with High Sensitivity RNA Assay. RNA with RIN &gt;8 was used. A total of 1 μg of high-quality RNA was used as input for transcriptomic library preparation for short-read sequencing. Library preparation was carried out using the KAPA RNA HyperPrep Kit, which includes ribosomal RNA (rRNA) depletion with the KAPA RiboErase (Human, Mouse, Rat) reagent to remove rRNA. Following rRNA depletion, the RNA was fragmented by heating to 94°C for 6 minutes. First-strand cDNA synthesis was then performed using random primers, followed by reverse transcription and second-strand synthesis. The resulting cDNA was subjected to end repair to generate blunt-ended fragments suitable for adapter ligation. KAPA Unique Dual-Indexed Adapters at a concentration of 3 μM were added to the cDNA fragments, and the ligation products were enriched through 6 cycles of PCR amplification. The amplified libraries were size-selected and purified using magnetic bead-based cleanup to remove undesired fragments and obtain the appropriate fragment size range. Library quality and size distribution were assessed using a TapeStation High Sensitivity D1000 Assay, and library concentration was quantified using the Qubit 1x dsDNA HS Kit to ensure optimal input for sequencing. Libraries were sequenced on an Illumina NextSeq 550 platform, generating paired-end reads.","Worms were flash-frozen in liquid nitrogen. Total RNA was isolated using the mirVana™ miRNA Isolation Kit  following the manufacturer’s protocol, then further fractionated into long (&gt;200 nt) and short (&lt; 200 nt) RNA fractions using the same kit. Short RNA fractions were then treated with TURBO™ DNase as per the provided instructions. The concentration of the purified RNA samples was quantified using a Qubit 4 Fluorometer and the Qubit™ RNA Broad Range (BR) Assay Kit. RNA quality was assessed using the Agilent TapeStation system with High Sensitivity RNA Assay. 275 ng RNA was used for library preparation with TruSeq Small RNA Library Preparation Kit (Illumina). The protocol included adapter ligation, reverse transcription and PCR amplification. The amplified cDNA was subsequently gel-purified. A band ranging from about 140 to 210 bp was cut, recovered and concentrated via ethanol precipitation. Library quality and size distribution were assessed using a TapeStation High Sensitivity D1000 Assay, and library concentration was quantified using the Qubit 1x dsDNA HS Kit  to ensure optimal input for sequencing. Libraries were sequenced on an Illumina NextSeq 550 platform, generating single-end reads.","Worms were flash-frozen in liquid nitrogen. Total RNA was isolated using the mirVana™ miRNA Isolation Kit following the manufacturer’s protocol, and treated with TURBO™ DNase as per the provided instructions. The concentration of the purified RNA samples was quantified using a Qubit 4 Fluorometer and the Qubit™ RNA Broad Range (BR) Assay Kit. RNA quality was assessed using the Agilent TapeStation system with High Sensitivity RNA Assay. RNA with RIN &gt;8 was used. Ribosomal RNA (rRNA) depletion was done with the riboPOOL kit. The sample was mixed with 1% spike-in, SIRV Set-4 (Iso Mix E0 / ERCC / Long SIRVs). Library was prepared and sequenced at Macrogen, Inc., South Korea.","Worms were flash-frozen in liquid nitrogen. Total RNA was isolated using the mirVana™ miRNA Isolation Kit following the manufacturer’s protocol, and treated with TURBO™ DNase as per the provided instructions. The concentration of the purified RNA samples was quantified using a Qubit 4 Fluorometer and the Qubit™ RNA Broad Range (BR) Assay Kit. RNA quality was assessed using the Agilent TapeStation system with High Sensitivity RNA Assay. RNA with RIN &gt;8 was used. Ribosomal RNA (rRNA) depletion was done with the riboPOOL kit. The sample was mixed with 1% spike-in, SIRV Set-4 (Iso Mix E0 / ERCC / Long SIRVs). Library was prepared and sequenced at Macrogen, Inc., South Korea.","Worms were flash-frozen in liquid nitrogen. Total RNA was isolated using the mirVana™ miRNA Isolation Kit following the manufacturer’s protocol, and treated with TURBO™ DNase as per the provided instructions. The concentration of the purified RNA samples was quantified using a Qubit 4 Fluorometer and the Qubit™ RNA Broad Range (BR) Assay Kit. RNA quality was assessed using the Agilent TapeStation system with High Sensitivity RNA ScreenTapes. RNA with RIN &gt;8 was used. 100 ng DNase-treated total RNA was then mixed with 1% spike-in, SIRV Set-4 (Iso Mix E0 / ERCC / Long SIRVs). Library was prepared using Iso-Seq express 2.0 kit and Iso-Seq SMRTbell prep kit 3.0. Library quality and size distribution were assessed using a TapeStation High Sensitivity D5000 Assay, and library concentration was quantified using the Qubit 1x dsDNA HS Kit to ensure optimal input for sequencing. Libraries were sequenced on a PacBio Sequel IIe.","Worms were flash-frozen in liquid nitrogen. Total RNA was isolated using the mirVana™ miRNA Isolation Kit following the manufacturer’s protocol, and treated with TURBO™ DNase as per the provided instructions. The concentration of the purified RNA samples was quantified using a Qubit 4 Fluorometer and the Qubit™ RNA Broad Range (BR) Assay Kit. RNA quality was assessed using the Agilent TapeStation system with High Sensitivity RNA ScreenTapes. RNA with RIN &gt;8 was used. DNase-treated total RNA was then rRNA depleted with riboPOOL kit and mixed with 1% spike-in, SIRV Set-4 (Iso Mix E0 / ERCC / Long SIRVs). Library was prepared using Iso-Seq express 2.0 kit and Iso-Seq SMRTbell prep kit 3.0. Library quality and size distribution were assessed using a TapeStation High Sensitivity D5000 Assay, and library concentration was quantified using the Qubit 1x dsDNA HS Kit to ensure optimal input for sequencing. Libraries were sequenced on a PacBio Sequel IIe.","Whole worms were dissociated into single-cell suspensions and processed using the 10x Genomics Chromium Single Cell 3′ workflow to generate full-length cDNA. rRNA sequences were depleted from approximately 1 ng of cDNA using scDASH (CRISPR-Cas9 digestion) with two sgRNA mixes at 300 nM and two-fold Cas9 enzyme, using a 16S:18S:28S targeting ratio of 1:1:2; sgRNAs were prepared individually and pooled prior to scDASH treatment (pooling step performed earlier than in the reference scDASH workflow). Libraries were sequenced with a 25% PhiX spike-in on an Illumina NovaSeq 6000.","Dissociated worm cell suspensions were stained with DRAQ5 and Calcein AM and sorted on a BD FACS Aria Fusion. For each biological replicate, 50,000 events were collected from the D (neoblast-enriched) gate and 50,000 events from the non-D gate. RNA was isolated from the sorted fractions and small RNA libraries were prepared using the GenXPro small RNA-Seq Library Preparation Kit (3′ adapter ligation, 5′ adapter ligation, reverse transcription, indexed PCR amplification, SPRI bead purification, gel-free workflow). Libraries were sequenced single-end (150 cycles) on one NextSeq High Output flow cell."],["Transcriptomic","Transcriptomic","Transcriptomic","Transcriptomic","Transcriptomic","Transcriptomic","Transcriptomic","Transcriptomic"],["cDNA","cDNA","cDNA","cDNA","cDNA","cDNA","cDNA (10x Genomics 3′ GEX, Standard kit)","cDNA"],["Illumina NextSeq 550","Illumina NextSeq 550","Illumina Novaseq 6000","PacBio Sequel II","PacBio Sequel IIe","PacBio Sequel IIe","10x Genomics Chromium Controller; Illumina NovaSeq 6000","Illumina NextSeq 550"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th>BioProject<\/th>\n      <th>Title<\/th>\n      <th>Organism<\/th>\n      <th>Experiment type<\/th>\n      <th>Summary<\/th>\n      <th>Treatment protocol<\/th>\n      <th>Growth protocol<\/th>\n      <th>Extracted molecule<\/th>\n      <th>Extraction protocol<\/th>\n      <th>Library source<\/th>\n      <th>Library selection<\/th>\n      <th>Instrument model<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"pageLength":25,"scrollX":true,"scrollY":520,"scroller":true,"dom":"Bfrtip","buttons":["copy","csv","excel"],"searchHighlight":true,"columnDefs":[{"name":"BioProject","targets":0},{"name":"Title","targets":1},{"name":"Organism","targets":2},{"name":"Experiment type","targets":3},{"name":"Summary","targets":4},{"name":"Treatment protocol","targets":5},{"name":"Growth protocol","targets":6},{"name":"Extracted molecule","targets":7},{"name":"Extraction protocol","targets":8},{"name":"Library source","targets":9},{"name":"Library selection","targets":10},{"name":"Instrument model","targets":11}],"order":[],"autoWidth":false,"orderClasses":false},"selection":{"mode":"multiple","selected":null,"target":"row","selectable":null}},"evals":[],"jsHooks":[]}</script>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
