<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Tables</title>

<script src="site_libs/header-attrs-2.30/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>
<link href="site_libs/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet" />
<script src="site_libs/htmlwidgets-1.6.4/htmlwidgets.js"></script>
<link href="site_libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet" />
<script src="site_libs/datatables-binding-0.34.0/datatables.js"></script>
<link href="site_libs/dt-core-1.13.6/css/jquery.dataTables.min.css" rel="stylesheet" />
<link href="site_libs/dt-core-1.13.6/css/jquery.dataTables.extra.css" rel="stylesheet" />
<script src="site_libs/dt-core-1.13.6/js/jquery.dataTables.min.js"></script>
<link href="site_libs/crosstalk-1.2.2/css/crosstalk.min.css" rel="stylesheet" />
<script src="site_libs/crosstalk-1.2.2/js/crosstalk.min.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Supplementary</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="umap.html">UMAP</a>
</li>
<li>
  <a href="figures.html">Figures</a>
</li>
<li>
  <a href="tables.html">Tables</a>
</li>
<li>
  <a href="scripts.html">Scripts</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Tables</h1>

</div>


<p>Click <strong>Open preview</strong> to view an interactive HTML
rendering of the XLSX file in the browser. Use <strong>Download</strong>
to save the original <code>.xlsx</code> file.</p>
<div class="datatables html-widget html-fill-item" id="htmlwidget-838241e2f58034ad9040" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-838241e2f58034ad9040">{"x":{"filter":"none","vertical":false,"data":[["Table S1.xlsx","Table S2.xlsx","Table S3.xlsx","Table S4.xlsx","Table S5.xlsx","Table S6.xlsx","Table S7.xlsx","Table S8.xlsx"],["<details><summary>Show<\/summary>Dataset-level metadata, including NCBI accessions, where available, treatment and extraction protocols, read configuration, and platform details, are provided in Supplementary Table S1. Bulk RNA-seq could detect strong whole-worm transcriptional responses, but it reported an average signal from mixed cell types and could obscure lineage-restricted changes, particularly when the relevant populations were rare (X. Li and Wang 2021). Thus, in the subsequent stage of the study, the impact of Smed ELAC2 silencing was examined at the single-cell resolution. In this analysis, datasets from WT, GFP mock, and Smed ELAC2 KD worms were used, encompassing samples collected at early stages of regeneration (0, 16, 24, and 72 hpa) (Supplementary Table S1). The 72 hpa time point corresponded to one of the bulk RNA-seq sampling points (3 dpa). The generation of single-cell resolution gene expression data was intended to be subsequently integrated with information on sncRNA abundance and predicted functions, in order to gain insight into their roles in regulating cell lineages. Pre-processing included read mapping, cell and feature QC, mitochondrial content filtering, doublet control, construction of 4 per-time point separate Seurat objects, containing all 3 conditions, and initial clustering. These steps were completed by Dr Małgorzata Marszałek-Zeńczak, and all analyses reported below start from these pre-processed objects (Section 3.9).<\/details>","<details><summary>Show<\/summary>To identify which sncRNA species were affected by Smed ELAC2 silencing and to guide subsequent analyses, differential accumulation was next evaluated between Smed ELAC2 KD and controls. Differential accumulation was assessed for sncRNA species, defined here as unique nucleotide sequences observed in the libraries. Differential accumulation was tested as described in Section 3.5 using DESeq2 with Benjamini–Hochberg correction; significance was defined as adjusted p ≤ 0.05 with |log2 fold change| ≥ 2, and sncRNAs were required to meet minimal abundance criteria (CPM &gt; 10 in at least two replicates). To reduce potential confounding from nonspecific RNA or injection, species showing control-specific changes (GFP mock versus WT) were excluded, leaving only Smed ELAC2 KD-dependent alterations (Suplementary Table S2).<\/details>","<details><summary>Show<\/summary>To relate sncRNA changes to co-occurring effects at the mRNA level, long RNA-seq data from the same regenerative time points were analyzed in parallel. Differential gene expression testing was performed using a conservative strategy intended to attribute changes specifically to Smed ELAC2 KD rather than injection-related, nonspecific RNA or background effects. Genes were required to be significantly altered in Smed ELAC2 KD relative to wild type while remaining stable between the 2 control groups, and additional abundance filtering and a minimum effect-size threshold (|log2 fold change| ≥ 0.57 were applied (Section 3.4). Under these criteria, the number of differentially expressed genes (DEGs) was low, with 19 upregulated and 27 downregulated genes at 3 dpa, increasing to 25 upregulated and 58 downregulated genes at 5 dpa (Supplementary Table S3). A standard step in transcriptomic analysis is functional summarization of deregulated genes by Gene Ontology (GO) assignment, covering biological processes, molecular functions, and cellular components, followed by enrichment analysis. In S. mediterranea, GO coverage is incomplete for many loci, and therefore a combined GO annotation set was constructed to support enrichment analysis. First, GO assignments available in PlanMine were retrieved for genes with existing planarian annotations. For genes lacking PlanMine GO terms, additional GO assignments were transferred using homology-based evidence generated during annotation, including best-supported SwissProt matches and domain-based mappings derived from InterProScan and Pfam, and these sources were merged into a unified gene-to-GO table (Supplementary Table S3) used for enrichment testing (Sections 3.3 and 3.4). Using this approach, GO terms were identified for 18,522 of the 25,264 genes.<\/details>","<details><summary>Show<\/summary>Cluster annotation, i.e., assigning biological identities to clusters, is generally challenging and becomes particularly difficult in non-classical model organisms, due to the lack of dedicated reference databases. To facilitate this process, a comprehensive and manually curated marker reference had to be constructed. To this end, a planarian-specific marker reference table (Supplementary Table S4) was prepared by collecting lineage and subtype markers from published planarian scRNA-seq atlases and lineage studies, recording the original gene names and sources, and then mapping each marker to the gene identifiers used in this dissertation (the schMed_Pn annotation). For each entry, a general lineage label and a detailed population label were assigned, together with both a human-readable marker name and the corresponding SMESG gene ID used in the pre-processed Seurat object. Duplicates and naming variants were harmonized so that each marker was represented consistently across analyses, and supporting references were retained in the table to enable manual traceability.<\/details>","<details><summary>Show<\/summary>This approach yielded a final atlas of 74,403 cells and 36,912 detected genes, resolving 57 biologically coherent populations spanning multiple neoblast states (σ, ζ, γ, ν and lineage-biased subclasses), epidermal progenitors and mature epidermal layers (including multiciliated states), intestinal phagocytes and goblet cells, protonephridial flame and tubule lineages with precursors, ECM-producing and pole/PCG muscle states, glia, diverse neuronal classes (including cholinergic, glutamatergic, catecholaminergic, neuropeptidergic, and sensory subtypes), eye lineages, and multiple parenchymal subtypes (Supplementary Table S5; Figure 25). PGRN⁺ parenchymal and abraçada cells were extremely similar and adjacent in transcriptional space. Both populations consistently showed PGRN and CTSL2 expression among their common markers, and the difference largely reflected relative detection frequency: PGRN was detected in a similar fraction of cells across both groups (means ~81–82%; Supplementary Table S5) during manual verification of final cell annotation labeling, whereas CTSL2 was detected in a smaller fraction of cells in PGRN⁺ parenchymal clusters (mean ~48%) than in abraçada clusters (mean ~82%), consistent with a CTSL2-centered definition for abraçada cells (DOI: 10.1101/2025.08.01.668211). In addition, phagocyte clusters were distinguished by near-uniform PSAP detection (mean ~87%) together with additional co-markers that were less frequent or less dominant in the PGRN⁺ parenchymal and abraçada cells.. Thus, PSAP primarily reported a shared lysosome-rich program across these populations, but phagocytes in this dataset were best defined by PSAP in combination with the broader phagocyte marker set rather than by PSAP alone.This interpretation was mechanistically plausible because, in other animals, both progranulin (encoded by PGRN) and prosaposin (encoded by PSAP) are lysosome-directed proteins that physically interact and support lysosomal trafficking, and progranulin can be proteolytically processed in lysosomes by cathepsin L (Zhou et al. 2015; C. W. Lee et al. 2017).<\/details>","<details><summary>Show<\/summary>Cell population-resolved DE was computed using the atlas labels described in Section 4.4.2.1 and the full details were provided in Section 3.15. Bulk-style replicate-based modeling (for example DESeq2) was not applicable here because scRNA-seq was available as a single biological replicate per condition and time point, and cell-level testing could be biased when between-replicate variation cannot be estimated (Squair et al. 2021). Accordingly, results were generated as within-sample, population-resolved signatures supported by effect size and by concordance across the set of time-matched control samples (Supplementary Table S6). A gene was treated as differentially expressed in response to Smed ELAC2 silencing in a given population and time point if it was significant in Smed ELAC2 KD vs WT and was also not significant in GFP mock vs WT in the same direction (Section 3.15).<\/details>","<details><summary>Show<\/summary>To identify Smed ELAC2 KD-specific effects, a Δ (delta) connectivity measure was used between Smed ELAC2 KD and controls per edge and time point. For each edge, Δ can be positive or negative. A positive Δ means the connection was stronger in Smed ELAC2 KD than expected from controls, while a negative Δ means it was weaker. Uncertainty was assessed by repeated resampling of cells, which provided confidence intervals for Δ (Section 3.16; Supplementary Table S7). For the family-level visualization, edges with very small absolute changes were ignored, specifically, changes with |Δ| &lt; 0.003 were treated as negligible for the summary plot (Figure 32).<\/details>","<details><summary>Show<\/summary>A low-input sncRNA library (GenXPro) was analyzed from the X1 and X2/Xins cells for WT, GFP mock, and Smed ELAC2 KD worms as described in Section 3.17. Reads were first assigned to broad RNA classes to evaluate overall library composition (Figure 34). Because rRFs reads dominated the libraries, they were excluded prior to differential testing. sncRNAs were retained for downstream comparisons only if they had at least 10 total reads summed across all samples (Suplementary Table S8). Given the lower information content of the fractionated libraries compared with the bulk sncRNA-seq experiment (Section 4.2.5), tRF reads were summarized at the tRNA isoacceptors level rather than at the finer resolution of parental tRNA subregions.<\/details>"],["<a href=\"table_Table_S1.html\">Open preview<\/a>","<a href=\"table_Table_S2.html\">Open preview<\/a>","<a href=\"table_Table_S3.html\">Open preview<\/a>","<a href=\"table_Table_S4.html\">Open preview<\/a>","<a href=\"table_Table_S5.html\">Open preview<\/a>","<a href=\"table_Table_S6.html\">Open preview<\/a>","<a href=\"table_Table_S7.html\">Open preview<\/a>","<a href=\"table_Table_S8.html\">Open preview<\/a>"],["<a href=\"tables/Table%20S1.xlsx\" download>Download<\/a>","<a href=\"tables/Table%20S2.xlsx\" download>Download<\/a>","<a href=\"tables/Table%20S3.xlsx\" download>Download<\/a>","<a href=\"tables/Table%20S4.xlsx\" download>Download<\/a>","<a href=\"tables/Table%20S5.xlsx\" download>Download<\/a>","<a href=\"tables/Table%20S6.xlsx\" download>Download<\/a>","<a href=\"tables/Table%20S7.xlsx\" download>Download<\/a>","<a href=\"tables/Table%20S8.xlsx\" download>Download<\/a>"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th>File<\/th>\n      <th>Mention<\/th>\n      <th>Open<\/th>\n      <th>Download<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"pageLength":25,"scrollX":true,"dom":"Bfrtip","buttons":["copy","csv"],"columnDefs":[{"name":"File","targets":0},{"name":"Mention","targets":1},{"name":"Open","targets":2},{"name":"Download","targets":3}],"order":[],"autoWidth":false,"orderClasses":false},"selection":{"mode":"multiple","selected":null,"target":"row","selectable":null}},"evals":[],"jsHooks":[]}</script>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
